cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/SourceControlManager.java

```diff
@@ -0,0 +1,253 @@
+    // Check if default branch exists.
+    String foundDefaultBranch =
+      refs.stream().map(Ref::getName).filter(defaultBranchRefName::equals).findAny().orElse(null);
+    if (foundDefaultBranch != null) {
+      return null;
+    }
+    return new RepositoryValidationFailure(String.format(
+      "Default branch not found in remote repository. Ensure ref '%s' already exists.",
+      defaultBranchRefName));
+  }
+
+  /**
+   * Validated the configuration provided to this manager.
+   *
+   * @return a {@link List} of {@link  RepositoryValidationFailure}s occurred during validation.
+   */
+  public RepositoryValidationFailure validateConfig() throws IOException {
+    // Validate existence of secret in Secure Store.
+    RepositoryConfig config = context.getRepositoryConfig();
+    LOG.info("Checking existence of secret in secure store: " + config.getAuth().getTokenName());
+    try {
+      context.getStore().get(context.getNamespaceId().getNamespace(), config.getAuth().getTokenName()).get();
+    } catch (Exception e) {
+      LOG.error("Failed to get auth credentials from secure store.", e);
+      return new RepositoryValidationFailure("Failed to get password/token from secure store.");
+    }
+
+    // Try fetching heads in the remote repository.
+    FileRepository localRepo = null;
+    deleteIfRequired(validationRepoDirectory);
+    try {
+      localRepo = new FileRepository(validationRepoDirectory);
+    } catch (IOException e) {
+      throw new IOException(String.format("Failed to create local git repository at path '%s'",
+                                          validationRepoDirectory.getAbsolutePath()), e);
+    }
+
+    LOG.info("Created local repo in directory for validation: " + localRepoDirectory.getAbsolutePath());
+    try (Git git = new Git(localRepo)) {
+      LsRemoteCommand cmd = git.lsRemote().setRemote(config.getLink()).setHeads(true).setTags(false);
+      cmd = getAuthenticatedCommandWithTimeout(cmd);
+      RepositoryValidationFailure failure = validateDefaultBranch(cmd.call());
+      if (failure != null) {
+        return failure;
+      }
+    } catch (TransportException e) {
+      LOG.error("Failed to list remotes in remote repository.", e);
+      return new RepositoryValidationFailure(
+        "Failed to connect with repository. Authentication credentials invalid or" + "the repository may not exist.");
+    } catch (Exception e) {
+      LOG.error("Failed to list remotes in remote repository.", e);
+      return new RepositoryValidationFailure(e.getMessage());
+    }
+    deleteIfRequired(validationRepoDirectory);
+    return null;
+  }
+
+  // Validated whether git is initialised.
+  private void validateInitialised() {
+    if (git == null) {
+      throw new IllegalStateException("Initialise source control manager before performing operation.");
+    }
+  }
+
+  public void push(CommitMeta meta) {
+    validateInitialised();
+    // TODO: Check status and skip if not required. Add + commit + push.
```

Comment:
nit: better to add JIRA number at TODOs

Reply:
I've removed methods without implementations, will add them in later PRs.